### ISSUE TITLE
"utils.Scope" introduces cyclic references

### ISSUE THREAD

**Issue by d-maurer** [malthe/chameleon#301]
`utils.Scope.new` uses `inst.set_global = inst.__setitem__`. This introduces a cyclic structure which needs to be reclaimed by full garbage collection (rather than via reference counting).

I suggest to avoid this by introducing an attribute `root` which is `None` for the root and otherwise the root scope and adjust the definitions of `set_global` and `copy` accordingly.

Note, the values set by `set_global` are not visible in non-root scopes (which is likely not what you want).

**Comment by malthe** [comment]
@d-maurer can you take a look at this fix?

**Comment by d-maurer** [comment]
Malthe Borch wrote at 2020-3-18 03:16 -0700:
>@d-maurer can you take a look at this fix?

Looks good!

**Comment by malthe** [comment]
The test is kind of sketchy. The return value of `gc.collect()` is both freed objects and dangling objects. But since I check immediately prior to deleting the speciment, I feel relatively confident about it.

On PyPy, there's no such return value but that's okay.

**Comment by d-maurer** [comment]
@malthe Could you make a new release please? It would allow to clean up "https://github.com/zopefoundation/Zope/pull/807".

As a side note: working on the above PR, I noticed that your implementation of "set_global" behaves differently from that of `zope.tales`: there, setting a global variable affects all scopes; your "set_global" affects only the global scope and local scopes which do not have a local definition for the variable. I think, your implementation is correct. I just want to let you know about the difference.

**Comment by malthe** [comment]
I use the global scope as a fallback to local lookup. I'm not sure how `zope.tales` has it exactly to be honest but if you have a specific reference then that might help in documenting or fixing any difference.

**Comment by d-maurer** [comment]
Malthe Borch wrote at 2020-3-25 22:55 -0700:
>I use the global scope as a fallback to local lookup. I'm not sure how `zope.tales` has it exactly to be honest but if you have a specific reference then that might help in documenting or fixing any difference.

The equivalent of `chameleon`s `Scope` in `zope.tales`
is `zope.tales.tales.Context`.

It has attributes `vars` (the currently used `variable -> value`
mapping) and `_vars_stack` (a stack of `vars` mappings).
Its `setGlobal` looks like:
```
    def setGlobal(self, name, value):
        for vars in self._vars_stack:
            vars[name] = value
```

**Comment by malthe** [comment]
That seems pretty expensive, but I guess people mostly set globals in the initial scope.

**Comment by d-maurer** [comment]
Malthe Borch wrote at 2020-3-26 00:01 -0700:
>That seems pretty expensive, but I guess people mostly set globals in the initial scope.

As I wrote: I think your implementation is good. In my feeling, it is
strange that a global setting can override a local one - even
if the global setting happened after the local one: I would be really
surprised if in a situation:
```
  <... tal:define="x ...">
     <...>
        ... something involving macros ...
     </...>
     <... tal:content:"...x..." />
     ...
```
the `x` in `tal:content` would not have the value from the `tal:define`
(because one of the macros had globally set the value for `x`).

**Comment by malthe** [comment]
@d-maurer I have released 3.7.0 today with this issue fixed.

**Comment by d-maurer** [comment]
Malthe Borch wrote at 2020-3-26 10:10 -0700:
>@d-maurer I have released 3.7.0 today with this issue fixed.

Thank you!
